### PR TITLE
Add Gazebo multi-robot simulator and dependencies to nixpkgs

### PR DESCRIPTION
**Pull request by therealpxc**
This is my first serious attempt at contributing, but I have plans for adding other packages as well. Please let me know if any changes are needed.

### CODE REVIEW

**Comment by nico202** [comment]
(Compiled it and is working, thanks) Travis fails because licence apache is called asl20

**Comment by therealpxc** [comment]
Awesome. I'll update the license setting.

**Comment by nico202** [comment]
@therealpxc do you also use yarp + gazebo-yarp-plugins? I've started porting [yarp](https://gist.github.com/nico202/4d35ed4c683c557b7dd0) but [libace](http://www.cs.wustl.edu/~schmidt/ACE/) is quite a mess, don't know where to start

**Comment by therealpxc** [comment]
I'm afraid I'm not familiar with yarp. We've not yet gotten very deep into a specific Gazebo workflow at my workplace. Right now we're just trying to get Nixy builds and continuous integration for PX4 and Gazebo. We're mostly looking to simulate quadcopters for now. What's yarp? I'd be happy to work on getting it packaged after we get initial support for some of this stuff, especially it might apply to my use case.

PS: I updated the meta information to be correct. :-)

**Comment by therealpxc** [comment]
We might have friends in those interested in #11713 as well. I'm not much interested in leading, but I'd be happy to work on a 'Nix robotics packages maintainers' team.

**Comment by nico202** [comment]
On the gist I linked (click on yarp on my previuous post), I already build it without ace. They say the function are limited though.

**Comment by therealpxc** [comment]
@nico202 What about [the way Homebrew does it](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/ace.rb)?

**Comment by nico202** [comment]
I'll try tomorrow (I think you need to rebase your commit. Travis failing because of this)

**Comment by therealpxc** [comment]
It's not continuing to fail, as far as I can tell. I just had to add that second commit. Am I mistaken?

**Comment by nico202** [comment]
yes my fault

File: pkgs/applications/science/misc/gazebo/default.nix [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    ++ optional withBulletEngineSupport bullet
+    ++ optional withIgnitionTransport ignition-transport
+    ++ optional withModelEditorSchematicView graphviz
+    ++ optional withDigitalElevationTerrainsSupport gdal
+    ++ optional withConstructiveSolidGeometrySupport gts
+    ++ optional withHdf5Support hdf5;
+
+  meta = with stdenv.lib; {
+    homepage = http://gazebosim.org/;
+    description = "Gazebo is a multi-robot simulator for outdoor environments";
```
Comment (jagajaga) [inline pkgs/applications/science/misc/gazebo/default.nix]: Please omit pkg name in the description.

**Comment by therealpxc** [comment]
Updated the descriptions for Gazebo and TinyXML-2. Thanks

**Comment by acowley** [comment]
Is KDE really required? I'd be interested in making this installable with Nix on OS X.

**Comment by therealpxc** [comment]
It only uses Qt. Qt is in the namespace kde4. Maybe I could change that line at the top.

I am also interested in getting this to work on OS X, but at present I don't have any hardware for testing on Darwin.

**Comment by therealpxc** [comment]
Are any further changes required?

File: pkgs/development/libraries/sdformat/default.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      sha256 = srchash-sha256;
+  };
+
+  inherit name;
+
+  enableParallelBuilding = true;
+  buildInputs = [
+    cmake boost ruby ignition-math2 tinyxml
+  ];
+}
```
Comment (fpletz) [inline pkgs/development/libraries/sdformat/default.nix]: Could you please add some meta attributes for this library?

**Comment by fpletz** [comment]
Good work considering this is your first contribution to NixOS! :beers:

If you add the meta attributes mentioned in my comment above, I'll merge.

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -2011,6 +2017,16 @@ let
 
   ifuse = callPackage ../tools/filesystems/ifuse/default.nix { };
 
+  ignition-math = callPackage ../development/libraries/ignition-math { };
+  
+  ignition-math2 = ignition-math;
```
Comment (the-kenny) [inline pkgs/top-level/all-packages.nix]: Any reason for this alias?

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -2011,6 +2017,16 @@ let
...
+
+  ignition-transport0 = callPackage ../development/libraries/ignition-transport/0.9.0.nix { };
```
Comment (the-kenny) [inline pkgs/top-level/all-packages.nix]: Why are both `0` and `1` packaged?

**Comment by therealpxc** [comment]
@the-kenny Because I want to package Gazebo for development as well as use, and future versions are being built upon the new release, which has just come out. We need to have a fully functional, ready-to-use version and also keep up with development.

At the time that I made ignition-math-2, it wasn't clear whether I would need any versions earlier than 2.x. The OSRF website is a bit out of date and obscures the current releases of the ignition libraries, so ignition-math-2 doesn't look like the current release at first glance. There's also a 1.x release of Ignition math, and the naming difference is reflected upstream. (Their binary release tarballs are called like ignition-math-12345.tar.gz, and then ignition-math2-45678.tar.gz, etc.)

I also don't yet know whether some other tools I want to package will depend on different versions of the Ignition libraries, since they're intended for general use.

**Comment by therealpxc** [comment]
@fpletz done! :-D

**Comment by fpletz** [comment]
Oh, sorry, one more thing that hit me after reading @the-kenny's comments and your explanation:

Are these libraries intended for general use or are they just usable by or within gazebo? In the latter case it would probably make sense to give them their own attribute in `all-packages.nix` to group them like `gitAndTools` or `gst_all`. This keeps the global namespace clean. :wink: 

What do you think?

**Comment by therealpxc** [comment]
At present, I am using them only in Gazebo, but the libraries are for general use. Later, I have plans to package [ROS](http://www.ros.org/) & [ROS 2](http://www.ros2.org/). When I do so, I might want to put all of the Open-Source Robotics Foundation projects into a single namespace, because then we will have genuinely many of them. Can I determine the preferred layout in a new PR for those additions when I get to them? I'm not sure I can know yet what the best structure is.

**Comment by therealpxc** [comment]
@fpletz What's the right way to separate these packages out into their own namespace? If it's not much work I'll just do it.

**Comment by therealpxc** [comment]
I've implemented a few improvements, including that one recommended in the comments here by @fpletz. I've made them in a [new PR](https://github.com/NixOS/nixpkgs/pull/14211) here (https://github.com/NixOS/nixpkgs/pull/14211), for reasons explained there.

**Pull request closed without merging**